eth/gasprice/feehistory.go

```diff
@@ -0,0 +1,251 @@
+// fills in the rest of the fields.
+func (f *Oracle) processBlock(bf *blockFees, percentiles []float64) {
+	chainconfig := f.backend.ChainConfig()
+	if bf.baseFee = bf.header.BaseFee; bf.baseFee == nil {
+		bf.baseFee = new(big.Int)
+	}
+	if chainconfig.IsLondon(big.NewInt(int64(bf.blockNumber + 1))) {
+		bf.nextBaseFee = misc.CalcBaseFee(chainconfig, bf.header)
+	} else {
+		bf.nextBaseFee = new(big.Int)
+	}
+	bf.gasUsedRatio = float64(bf.header.GasUsed) / float64(bf.header.GasLimit)
+	if len(percentiles) == 0 {
+		// rewards were not requested, return null
+		return
+	}
+
+	bf.reward = make([]*big.Int, len(percentiles))
+	if len(bf.block.Transactions()) == 0 {
+		// return an all zero row if there are no transactions to gather data from
+		for i := range bf.reward {
+			bf.reward[i] = new(big.Int)
+		}
+		return
+	}
+	txs := make([]*types.Transaction, len(bf.block.Transactions()))
+	copy(txs, bf.block.Transactions())
+	sorter := newSorter(txs, bf.block.BaseFee())
+	sort.Sort(sorter)
+
+	var txIndex int
+	sumGasUsed := txs[0].Gas()
+
+	for i, p := range percentiles {
+		thresholdGasUsed := uint64(float64(bf.block.GasUsed()) * p / 100)
+		for sumGasUsed < thresholdGasUsed && txIndex < len(bf.block.Transactions())-1 {
+			txIndex++
+			sumGasUsed += txs[txIndex].Gas()
+		}
+		bf.reward[i], _ = txs[txIndex].EffectiveGasTip(bf.block.BaseFee())
+	}
+}
+
+// FeeHistory returns data relevant for fee estimation based on the specified range of blocks.
+// The range can be specified either with absolute block numbers or ending with the latest
+// or pending block. Backends may or may not support gathering data from the pending block
+// or blocks older than a certain age (specified in maxHistory). The first block of the
+// actually processed range is returned to avoid ambiguity when parts of the requested range
+// are not available or when the head has changed during processing this request.
+// Three arrays are returned based on the processed blocks:
+// - reward: the requested percentiles of effective priority fees per gas of transactions in each
+//   block, sorted in ascending order and weighted by gas used.
+// - baseFee: base fee per gas in the given block
+// - gasUsedRatio: gasUsed/gasLimit in the given block
+// Note: baseFee includes the next block after the newest of the returned range, because this
+// value can be derived from the newest block.
+func (f *Oracle) FeeHistory(ctx context.Context, blockCount int, lastBlockNumber rpc.BlockNumber, rewardPercentiles []float64) (firstBlockNumber rpc.BlockNumber, reward [][]*big.Int, baseFee []*big.Int, gasUsedRatio []float64, err error) {
+	if blockCount < 1 {
+		// returning with no data and no error means there are no retrievable blocks
+		return
+	}
+	if blockCount > maxBlockCount {
+		blockCount = maxBlockCount
+	}
+	for i, p := range rewardPercentiles {
+		if p < 0 || p > 100 || (i > 0 && p < rewardPercentiles[i-1]) {
+			return 0, nil, nil, nil, errInvalidPercentiles
+		}
+	}
+	var (
```

Comment:
We can simplify the thing like this 

```go

func (f *Oracle) resolveLastBlockNumber(last rpc.BlockNumber) (*types.Block, uint64, bool, error) {
	var (
		pending *types.Block
		noHead  bool
	)
	if last == rpc.PendingBlockNumber {
		pending, _ = f.backend.BlockByNumber(ctx, last)
		if pending != nil {
			return pending, pending.NumberU64(), false, nil
		}
		last, noHead = rpc.LatestBlockNumber, true
	}
	if last == rpc.LatestBlockNumber {
		latestHeader, err := f.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
		if err == nil {
			last = rpc.BlockNumber(latestHeader.Number.Uint64())
			return nil, last, noHead, nil
		} 
		return nil, 0, false, err
	}
	return nil, last, false, nil
}

```

The `headBlockNumber` actually is unnecessary. If pending is not nil and it's requested for processing later, we can use the `pending` field here for checking, instead of using `headBlockNumber`

And also the `latestHeader` is also kind of unnecessary since retrieve the header again is cheap because of the cache.

Reply:
And we can call this function like this. It's much cleaner.

```go

pending, last, noHead, err := f.resolveLastBlockNumber(lastBlockNumber)
	if err != nil {
		return 0, nil, nil, nil, err
	}
	if noHead {
		blockCount--
		if blockCount == 0 {
			return
		}
	}

```

Reply:
Actually I do need `headBlockNumber` to enforce `maxHistory` and avoid requesting future blocks. But factoring out the block number resolving logic is a great idea, I think it's a lot more readable now. Also you're right about `latestHeader`, saving it is not really imporant and not doing so makes the code simpler.

Reply:
The `headBlockNumber` is `last` returned by `resolveLastBlockNumber`

Reply:
That's not true if `lastBlockNumber` was an explicit block number different from the head. We still need the head though to apply `maxHistory` restriction and prevent future block requests. So I think it's better to also integrate these checks into `resolveBlockRange` that can also limit `blockCount` if necessary.